Call qnnpack's conv setup only if input pointer has changed.

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42840 Enable input pointer caching in XNNPACK integration.
* #42178 Change quantizer to account for input tensor's memory format.
* **#42008 Call qnnpack's conv setup only if input pointer has changed.**
* #42007 Refactor qconv to reduce allocations.
* #42006 Simple caching allocator for CPU.

Summary:
With caching allocator we have increased the likelihood of getting the
same input pointer. With that we can cache qnnpack operator and input
pointer and check if the input pointer is the same. If so we can skip
setup step.

Test Plan:
Ran one of the quantized models to observe
1. No pagefaults due to indirection buffer reallocation.
2. Much less time spent in indirection buffer population.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726973](https://our.internmc.facebook.com/intern/diff/D22726973)

kimishpatel committed:
Call qnnpack's conv setup only if input pointer has changed.

Summary:
With caching allocator we have increased the likelihood of getting the
same input pointer. With that we can cache qnnpack operator and input
pointer and check if the input pointer is the same. If so we can skip
setup step.

Test Plan:
Ran one of the quantized models to observe
1. No pagefaults due to indirection buffer reallocation.
2. Much less time spent in indirection buffer population.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

AshkanAliabadi commented:
I was thinking this might not be as effective as Caffe2's because the pointer is changing constantly in PyTorch.  I am curious if you noticed a performance boost.  If so, is it becuase the caching allocator?

kimishpatel commented:
> I was thinking this might not be as effective as Caffe2's because the pointer is changing constantly in PyTorch. I am curious if you noticed a performance boost. If so, is it becuase the caching allocator?

Yes. This is useless without caching allocator.

AshkanAliabadi commented:
Neat, it's very cool that we can get more mileage out of the allocator with
this.

On Fri, Jul 24, 2020, 10:35 AM Kimish Patel <notifications@github.com>
wrote:

> I was thinking this might not be as effective as Caffe2's because the
> pointer is changing constantly in PyTorch. I am curious if you noticed a
> performance boost. If so, is it becuase the caching allocator?
>
> Yes. This is useless without caching allocator.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/42008#issuecomment-663651220>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD24NKE53C4OK4BWFSJHDGLR5HA63ANCNFSM4PGYPQ6Q>
> .
>

kimishpatel commented:
Will also enable the same for xnnpack integration, later.

AshkanAliabadi commented:
Very cool, thanks.

On Fri, Jul 24, 2020, 10:42 AM Kimish Patel <notifications@github.com>
wrote:

> Will also enable the same for xnnpack integration, later.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/42008#issuecomment-663654177>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD24NKAACN5ZZVTW63QZ6O3R5HBYVANCNFSM4PGYPQ6Q>
> .
>

AshkanAliabadi commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/conv-run.cc:
```diff
@@ -325,20 +325,30 @@ enum pytorch_qnnp_status qnnpackConv(
+        input,
+        input_pixel_stride,
+        output,
+        output_pixel_stride,
```
Do output and output_pixel_stride need to be part of the `if` check at the top?

kimishpatel committed:
Update on "Call qnnpack's conv setup only if input pointer has changed."


Summary:
With caching allocator we have increased the likelihood of getting the
same input pointer. With that we can cache qnnpack operator and input
pointer and check if the input pointer is the same. If so we can skip
setup step.

Test Plan:
Ran one of the quantized models to observe
1. No pagefaults due to indirection buffer reallocation.
2. Much less time spent in indirection buffer population.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726973](https://our.internmc.facebook.com/intern/diff/D22726973)

[ghstack-poisoned]

kimishpatel committed:
Update on "Call qnnpack's conv setup only if input pointer has changed."


Summary:
With caching allocator we have increased the likelihood of getting the
same input pointer. With that we can cache qnnpack operator and input
pointer and check if the input pointer is the same. If so we can skip
setup step.

Test Plan:
Ran one of the quantized models to observe
1. No pagefaults due to indirection buffer reallocation.
2. Much less time spent in indirection buffer population.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726973](https://our.internmc.facebook.com/intern/diff/D22726973)

[ghstack-poisoned]

kimishpatel committed:
Update on "Call qnnpack's conv setup only if input pointer has changed."


Summary:
With caching allocator we have increased the likelihood of getting the
same input pointer. With that we can cache qnnpack operator and input
pointer and check if the input pointer is the same. If so we can skip
setup step.

Test Plan:
Ran one of the quantized models to observe
1. No pagefaults due to indirection buffer reallocation.
2. Much less time spent in indirection buffer population.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726973](https://our.internmc.facebook.com/intern/diff/D22726973)

[ghstack-poisoned]

kimishpatel committed:
Update on "Call qnnpack's conv setup only if input pointer has changed."


Summary:
With caching allocator we have increased the likelihood of getting the
same input pointer. With that we can cache qnnpack operator and input
pointer and check if the input pointer is the same. If so we can skip
setup step.

Test Plan:
Ran one of the quantized models to observe
1. No pagefaults due to indirection buffer reallocation.
2. Much less time spent in indirection buffer population.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726973](https://our.internmc.facebook.com/intern/diff/D22726973)

[ghstack-poisoned]

kimishpatel committed:
Update on "Call qnnpack's conv setup only if input pointer has changed."


Summary:
With caching allocator we have increased the likelihood of getting the
same input pointer. With that we can cache qnnpack operator and input
pointer and check if the input pointer is the same. If so we can skip
setup step.

Test Plan:
Ran one of the quantized models to observe
1. No pagefaults due to indirection buffer reallocation.
2. Much less time spent in indirection buffer population.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726973](https://our.internmc.facebook.com/intern/diff/D22726973)

[ghstack-poisoned]

kimishpatel committed:
Update on "Call qnnpack's conv setup only if input pointer has changed."


Summary:
With caching allocator we have increased the likelihood of getting the
same input pointer. With that we can cache qnnpack operator and input
pointer and check if the input pointer is the same. If so we can skip
setup step.

Test Plan:
Ran one of the quantized models to observe
1. No pagefaults due to indirection buffer reallocation.
2. Much less time spent in indirection buffer population.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726973](https://our.internmc.facebook.com/intern/diff/D22726973)

[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:
```diff
@@ -129,20 +129,26 @@ enum pytorch_qnnp_status qnnpackDeConv(
+        input,
+        input_pixel_stride,
+        output,
+        output_pixel_stride,
```
I've got a feeling output and output_pixel_stride should also be part of the check.  What if input has not changed but output or output_pixel_stride have?  Looking at the implementation of `pytorch_qnnp_setup_deconvolution2d_nhwc_q8` it seems these two fields also change the internal state of the operator.

kimishpatel replied:
The reason I did not include this is because while setup function does store it in operator, our integration of conv/deconv does not run via run_operator. Also output_pixel_stride is derived from conv_params_t which is setup once. So only thing that can potentially change is output pointer, but since we are not relying on that to run the operator, it seemed ok. 
As opposed to this in XNNPACK integration we do the check for output ptr as well since we cannot make any assumption on it.

kimishpatel committed:
Update on "Call qnnpack's conv setup only if input pointer has changed."


Summary:
With caching allocator we have increased the likelihood of getting the
same input pointer. With that we can cache qnnpack operator and input
pointer and check if the input pointer is the same. If so we can skip
setup step.

Test Plan:
Ran one of the quantized models to observe
1. No pagefaults due to indirection buffer reallocation.
2. Much less time spent in indirection buffer population.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726973](https://our.internmc.facebook.com/intern/diff/D22726973)

[ghstack-poisoned]

kimishpatel committed:
Update on "Call qnnpack's conv setup only if input pointer has changed."


Summary:
With caching allocator we have increased the likelihood of getting the
same input pointer. With that we can cache qnnpack operator and input
pointer and check if the input pointer is the same. If so we can skip
setup step.

Test Plan:
Ran one of the quantized models to observe
1. No pagefaults due to indirection buffer reallocation.
2. Much less time spent in indirection buffer population.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726973](https://our.internmc.facebook.com/intern/diff/D22726973)

[ghstack-poisoned]

kimishpatel committed:
Update on "Call qnnpack's conv setup only if input pointer has changed."


Summary:
With caching allocator we have increased the likelihood of getting the
same input pointer. With that we can cache qnnpack operator and input
pointer and check if the input pointer is the same. If so we can skip
setup step.

Test Plan:
Ran one of the quantized models to observe
1. No pagefaults due to indirection buffer reallocation.
2. Much less time spent in indirection buffer population.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726973](https://our.internmc.facebook.com/intern/diff/D22726973)

[ghstack-poisoned]

kimishpatel committed:
Update on "Call qnnpack's conv setup only if input pointer has changed."


Summary:
With caching allocator we have increased the likelihood of getting the
same input pointer. With that we can cache qnnpack operator and input
pointer and check if the input pointer is the same. If so we can skip
setup step.

Test Plan:
Ran one of the quantized models to observe
1. No pagefaults due to indirection buffer reallocation.
2. Much less time spent in indirection buffer population.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726973](https://our.internmc.facebook.com/intern/diff/D22726973)

[ghstack-poisoned]

kimishpatel committed:
Update on "Call qnnpack's conv setup only if input pointer has changed."


Summary:
With caching allocator we have increased the likelihood of getting the
same input pointer. With that we can cache qnnpack operator and input
pointer and check if the input pointer is the same. If so we can skip
setup step.

Test Plan:
Ran one of the quantized models to observe
1. No pagefaults due to indirection buffer reallocation.
2. Much less time spent in indirection buffer population.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726973](https://our.internmc.facebook.com/intern/diff/D22726973)

[ghstack-poisoned]

kimishpatel committed:
Update on "Call qnnpack's conv setup only if input pointer has changed."


Summary:
With caching allocator we have increased the likelihood of getting the
same input pointer. With that we can cache qnnpack operator and input
pointer and check if the input pointer is the same. If so we can skip
setup step.

Test Plan:
Ran one of the quantized models to observe
1. No pagefaults due to indirection buffer reallocation.
2. Much less time spent in indirection buffer population.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726973](https://our.internmc.facebook.com/intern/diff/D22726973)

[ghstack-poisoned]

kimishpatel committed:
Update on "Call qnnpack's conv setup only if input pointer has changed."


Summary:
With caching allocator we have increased the likelihood of getting the
same input pointer. With that we can cache qnnpack operator and input
pointer and check if the input pointer is the same. If so we can skip
setup step.

Test Plan:
Ran one of the quantized models to observe
1. No pagefaults due to indirection buffer reallocation.
2. Much less time spent in indirection buffer population.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726973](https://our.internmc.facebook.com/intern/diff/D22726973)

[ghstack-poisoned]

kimishpatel committed:
Update on "Call qnnpack's conv setup only if input pointer has changed."


Summary:
With caching allocator we have increased the likelihood of getting the
same input pointer. With that we can cache qnnpack operator and input
pointer and check if the input pointer is the same. If so we can skip
setup step.

Test Plan:
Ran one of the quantized models to observe
1. No pagefaults due to indirection buffer reallocation.
2. Much less time spent in indirection buffer population.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726973](https://our.internmc.facebook.com/intern/diff/D22726973)

[ghstack-poisoned]

kimishpatel committed:
Update on "Call qnnpack's conv setup only if input pointer has changed."


Summary:
With caching allocator we have increased the likelihood of getting the
same input pointer. With that we can cache qnnpack operator and input
pointer and check if the input pointer is the same. If so we can skip
setup step.

Test Plan:
Ran one of the quantized models to observe
1. No pagefaults due to indirection buffer reallocation.
2. Much less time spent in indirection buffer population.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726973](https://our.internmc.facebook.com/intern/diff/D22726973)

[ghstack-poisoned]

kimishpatel committed:
Update on "Call qnnpack's conv setup only if input pointer has changed."


Summary:
With caching allocator we have increased the likelihood of getting the
same input pointer. With that we can cache qnnpack operator and input
pointer and check if the input pointer is the same. If so we can skip
setup step.

Test Plan:
Ran one of the quantized models to observe
1. No pagefaults due to indirection buffer reallocation.
2. Much less time spent in indirection buffer population.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726973](https://our.internmc.facebook.com/intern/diff/D22726973)

[ghstack-poisoned]

kimishpatel committed:
Update on "Call qnnpack's conv setup only if input pointer has changed."


Summary:
With caching allocator we have increased the likelihood of getting the
same input pointer. With that we can cache qnnpack operator and input
pointer and check if the input pointer is the same. If so we can skip
setup step.

Test Plan:
Ran one of the quantized models to observe
1. No pagefaults due to indirection buffer reallocation.
2. Much less time spent in indirection buffer population.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726973](https://our.internmc.facebook.com/intern/diff/D22726973)

[ghstack-poisoned]

kimishpatel committed:
Update on "Call qnnpack's conv setup only if input pointer has changed."


Summary:
With caching allocator we have increased the likelihood of getting the
same input pointer. With that we can cache qnnpack operator and input
pointer and check if the input pointer is the same. If so we can skip
setup step.

Test Plan:
Ran one of the quantized models to observe
1. No pagefaults due to indirection buffer reallocation.
2. Much less time spent in indirection buffer population.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726973](https://our.internmc.facebook.com/intern/diff/D22726973)

[ghstack-poisoned]

kimishpatel committed:
Update on "Call qnnpack's conv setup only if input pointer has changed."


Summary:
With caching allocator we have increased the likelihood of getting the
same input pointer. With that we can cache qnnpack operator and input
pointer and check if the input pointer is the same. If so we can skip
setup step.

Test Plan:
Ran one of the quantized models to observe
1. No pagefaults due to indirection buffer reallocation.
2. Much less time spent in indirection buffer population.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22726973](https://our.internmc.facebook.com/intern/diff/D22726973)

[ghstack-poisoned]
